CRM/Civisocial/OAuthProvider/Facebook.php

```diff
@@ -0,0 +1,256 @@
+      exit("Invalid request.");
+    }
+
+    // Make an API request to obtain Access Token
+    // GET params
+    $params = array(
+      'client_id' => $this->apiKey,
+      'client_secret' => $this->apiSecret,
+      'code' => CRM_Utils_Array::value('code', $_GET),
+      'redirect_uri' => $this->getCallbackUri($this->alias),
+    );
+
+    $response = $this->get('oauth/access_token', $params);
+    if (isset($response['error'])) {
+      exit($response['error']);
+    }
+
+    $this->token = CRM_Utils_Array::value('access_token', $response);
+
+    // Check if all basic perimissions have been granted
+    $deniedPermissions = $this->checkPermissions($this->getBasicPermissions());
+
+    if (!empty($deniedPermissions)) {
+      CRM_Utils_System::redirect($this->getLoginUri($deniedPermissions, TRUE));
+      // @todo:	It would be better if we inform first (eg. You need to provide
+      //			email to continue) and then provide a link to re-authorize
+    }
+
+    // Authentication is successful. Fetch user profile
+    $userProfile = array();
+    if ($this->isAuthorized()) {
+      $userProfile = $this->getUserProfile();
+    }
+    else {
+      // Start over
+      CRM_Utils_System::redirect($this->getLoginUri());
+    }
+
+    $facebookUserId = CRM_Utils_Array::value("id", $userProfile);
+    $this->login($this->alias, $this->token, $facebookUserId);
+
+    if (!CRM_Civisocial_BAO_CivisocialUser::socialUserExists($facebookUserId, $this->alias)) {
+      $user = array(
+        'first_name' => CRM_Utils_Array::value("name", $userProfile),
+        'last_name' => '',
+        'display_name' => CRM_Utils_Array::value("name", $userProfile),
+        'preffered_language' => CRM_Utils_Array::value("lang", $userProfile),
+        'gender' => NULL,
+        'email' => CRM_Utils_Array::value("email", $userProfile),
+        'contact_type' => 'Individual',
+      );
+
+      // Create contact
+      $contactId = CRM_Civisocial_BAO_CivisocialUser::createContact($user);
+
+      // Create social user
+      $socialUser = array(
+        'contact_id' => $contactId,
+        'social_user_id' => $facebookUserId,
+        'access_token' => $this->token,
+      // @todo: Rename oauth_object in table to oauth_secret?
+        'backend' => $this->alias,
+        'created_date' => time(), // @todo: Created Date not being recorded
+      );
+
+      CRM_Civisocial_BAO_CivisocialUser::create($socialUser);
+    }
+
+    CRM_Core_Session::setStatus(ts('Login via Facebook successful.'), ts('Login Successful'), 'success');
+    // @todo: Is status shown on public pages?
+    CRM_Utils_System::redirect($requestOrigin);
+  }
+
+  /**
+   * Check if the user is connected to Facebook and authorized.
+   * It can also be used to validate access tokens after setting one.
+   *
+   * @return bool
+   */
+  public function isAuthorized() {
+    if ($this->token && isset($this->userProfile)) {
+      return TRUE;
+    }
+    $response = $this->get('me');
```

Comment:
Any of the future interactions with any of the OAuthProviders after login will have following common steps:

1. Retrieve the saved access token from session/database.
2. Set the access token to Facebook/Googleplus/Twitter object.
3. Check the validity of the access token.
4. If the token is valid, retrieve logged in user info (at least ID.)

`GET /me` is the standard way to check if the access token is valid.  If the token is invalid the user information won't be returned; an Error will be thrown instead. Facebook REST API doesn't provide a node to check the validity of access token. Plus doing this would group Step 3 and 4.